### DISCUSSION TITLE
SQLAlchemy memory usage: Can I tell SQLAlchemy to NOT keep track of entities in the session?

### DISCUSSION THREAD

**Discussion by codeape2** [sqlalchemy/sqlalchemy#5894] [Usage Questions] [answered]
I some times use the SQLAlchemy ORM layer for read-only operations, typically for some sort of reporting task, something like:

    for item in session.query(Item):
        for relatedobject in item.relatedobjects:
            print(item.id, item.name, relatedobject.id, relatedobject.name)

In this case:

* I never modify the entities in the unit of work, and 
* The item -> relatedobject relationship is one-to-many (i.e. a relatedobject always belongs to a single item entity)

Question:
* Unless I am mistaken, all entities read from the db in the case above is tracked in memory by the SQLAlchemy session (for change tracking etc.)
* For some of these reporting loads, the python process uses a lot of memory due to this in-mem tracking (I believe)
* Given the case above, is there a way I can tell SQLAlchemy to not keep track of the entities in memory during the operation?

I do not want to use the SQLAlchemy expression layer, since I want to use the relationship functionality in the ORM layer.

(I asked this question on Stack Overflow as well: https://stackoverflow.com/questions/65918721/sqlalchemy-memory-usage-can-i-tell-sqlalchemy-to-not-keep-track-of-entities-in)

**Comment by zzzeek** [comment, marked as answer]
> 
>     * I never modify the entities in the unit of work, and

OK, this part is important

> 
> Question:
> 
>     * Unless I am mistaken, all entities read from the db in the case above is tracked in memory by the SQLAlchemy session (for change tracking etc.)

This part is true, but....

> 
>     * For some of these reporting loads, the python process uses a lot of memory due to this in-mem tracking (I believe)

this part is not true.  The Session only has weak references to your objects, as long as you aren't keeping pending, unflushed changes around.   Please see https://docs.sqlalchemy.org/en/13/orm/session_state_management.html#session-referencing-behavior .
